File: androidDemo/src/main/java/com/getyourmap/androidDemo/CurLocationHelper.java
package com.getyourmap.androidDemo;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.location.Criteria;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.util.Log;

import com.getyourmap.glmap.GLMapAnimation;
import com.getyourmap.glmap.GLMapDrawable;
import com.getyourmap.glmap.GLMapVectorCascadeStyle;
import com.getyourmap.glmap.GLMapVectorObject;
import com.getyourmap.glmap.GLMapView;
import com.getyourmap.glmap.MapPoint;

import java.util.List;

import static android.content.Context.LOCATION_SERVICE;

/** Created by destman on 6/1/17. */
class CurLocationHelper implements LocationListener {
  private static final long MIN_TIME_BW_UPDATES = 1000; // 1 second
  private static final long MIN_DISTANCE_CHANGE_FOR_UPDATES = 1; // 1 meter

  private GLMapDrawable userMovementImage, userLocationImage, accuracyCircle;
  private boolean isFollowLocationEnabled = false;
  private LocationManager locationManager;
  private Location lastLocation;
  private GLMapView mapView;

  CurLocationHelper(GLMapView mapView) {
    this.mapView = mapView;
  }

  void onDestroy() {
    if (locationManager != null) {
      locationManager.removeUpdates(this);
      locationManager = null;
    }
  }

  boolean initLocationManager(Activity activity) {
    if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION)
            != PackageManager.PERMISSION_GRANTED
        && ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION)
            != PackageManager.PERMISSION_GRANTED) return false;

    if (locationManager == null) {
      try {
        // Setup get location service
        locationManager = (LocationManager) activity.getSystemService(LOCATION_SERVICE);

        // Configure criteria
        Criteria criteria = new Criteria();
        criteria.setAccuracy(Criteria.ACCURACY_FINE);
        criteria.setSpeedRequired(true);
        criteria.setAltitudeRequired(true);
        criteria.setBearingRequired(true);
        criteria.setBearingAccuracy(Criteria.ACCURACY_HIGH);

        // Find the best location that currently have all location providers
        float bestAccuracy = Float.MAX_VALUE;
        long bestTime = Long.MAX_VALUE;
        lastLocation = null;
        List<String> matchingProviders = locationManager.getAllProviders();
        for (String provider : matchingProviders) {
          Location location = null;
          try {
            location = locationManager.getLastKnownLocation(provider);
          } catch (Exception e) {
            Log.e("CurLocationHelper", e.getLocalizedMessage());
          }
          if (location != null) {
            float accuracy = location.getAccuracy();
            long time = location.getTime();
            if ((time >= bestTime && accuracy < bestAccuracy)) {
              lastLocation = location;
              bestAccuracy = accuracy;
              bestTime = time;
            } else if (time < bestTime && bestAccuracy == Float.MAX_VALUE) {
              lastLocation = location;
              bestTime = time;
              bestAccuracy = accuracy;
            }
          }
        }
        // Update location to current best
        onLocationChanged(lastLocation);
        // Request location updates
        locationManager.requestLocationUpdates(
            MIN_TIME_BW_UPDATES,
            MIN_DISTANCE_CHANGE_FOR_UPDATES,
            criteria,
            this,
            activity.getMainLooper());
      } catch (Exception e) {
        locationManager = null;
      }
    }
    return true;
  }

  public Location getLastLocation() {
    return lastLocation;
  }

  public void setEnableFollowLocation(boolean enable) {
    this.isFollowLocationEnabled = enable;
    if (isFollowLocationEnabled) {
      onLocationChanged(lastLocation);
    }
  }

  @Override
  public void onLocationChanged(final Location location) {
    if (location == null) return;

    lastLocation = location;
    final MapPoint position =
        MapPoint.CreateFromGeoCoordinates(location.getLatitude(), location.getLongitude());
    if (isFollowLocationEnabled) {
      mapView.animate(
          new GLMapView.AnimateCallback() {
            @Override
            public void run(GLMapAnimation animation) {
              animation.flyToPoint(position);
            }
          });
    }

    // Create drawables if not exist and set initial positions.
    if (userLocationImage == null) {
      Bitmap locationImage =
          mapView.imageManager.open("DefaultStyle.bundle/circle-new.svgpb", 1, 0);
      if (locationImage != null) {
        userLocationImage = new GLMapDrawable(locationImage, 100);
        userLocationImage.setHidden(true);
        userLocationImage.setOffset(locationImage.getWidth() / 2, locationImage.getHeight() / 2);
        userLocationImage.setPosition(position);
        mapView.add(userLocationImage);
        locationImage.recycle();
      }
    }

    if (userMovementImage == null) {
      Bitmap movementImage = mapView.imageManager.open("DefaultStyle.bundle/arrow-new.svgpb", 1, 0);
      if (movementImage != null) {
        userMovementImage = new GLMapDrawable(movementImage, 100);
        userMovementImage.setHidden(true);
        userMovementImage.setOffset(movementImage.getWidth() / 2, movementImage.getHeight() / 2);
        userMovementImage.setRotatesWithMap(true);
        userMovementImage.setPosition(position);
        if (location.hasBearing()) userLocationImage.setAngle(-location.getBearing());
        mapView.add(userMovementImage);
        movementImage.recycle();
      }
    }

    // Select what image to display
    if (location.hasBearing()) {
      userMovementImage.setHidden(false);
      userLocationImage.setHidden(true);
    } else {
      userLocationImage.setHidden(false);
      userMovementImage.setHidden(true);
    }

    // Calculate radius of accuracy circle
    final float r = (float) mapView.convertMetersToInternal(location.getAccuracy());
    // If accuracy circle drawable not exits - create it and set initial position
    if (accuracyCircle == null) {
      final int pointCount = 100;
      // Use MapPoint to avoid distortions of circle
      MapPoint[] points = new MapPoint[pointCount];
      for (int i = 0; i < pointCount; i++) {
        double f = 2 * Math.PI * i / pointCount;
        // If radius of circle will be 1 only 2 points will be in final geometry (after
        // douglas-peucker)
        points[i] = new MapPoint(Math.sin(f) * 2048, Math.cos(f) * 2048);
      }
      GLMapVectorObject circle = GLMapVectorObject.createPolygon(new MapPoint[][] {points}, null);

      accuracyCircle = new GLMapDrawable(99);
      accuracyCircle.setTransformMode(GLMapDrawable.TransformMode.Custom);
      accuracyCircle.setPosition(position);
      accuracyCircle.setScale(r / 2048.0f);
      accuracyCircle.setVectorObject(
          mapView,
          circle,
          GLMapVectorCascadeStyle.createStyle(
              "area{layer:100; width:1pt; fill-color:#3D99FA26; color:#3D99FA26;}"),
          null);
      mapView.add(accuracyCircle);
    }

    mapView.animate(
        new GLMapView.AnimateCallback() {
          @Override
          public void run(GLMapAnimation animation) {
            animation.setTransition(GLMapAnimation.Linear);
            animation.setDuration(1);
            userMovementImage.setPosition(position);
            userLocationImage.setPosition(position);
            accuracyCircle.setPosition(position);
            accuracyCircle.setScale(r / 2048.0f);
            if (location.hasBearing()) userLocationImage.setAngle(-location.getBearing());
          }
        });
  }

  @Override
  public void onStatusChanged(String provider, int status, Bundle extras) {}

  @Override
  public void onProviderEnabled(String provider) {}

  @Override
  public void onProviderDisabled(String provider) {}
}

File: kotlinDemo/src/main/java/globus/demo/CurLocationHelper.kt
package globus.demo

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.location.Criteria
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.util.Log
import androidx.core.app.ActivityCompat
import globus.glmap.*
import kotlin.math.cos
import kotlin.math.sin

/** Created by destman on 6/1/17.  */
internal class CurLocationHelper(private val renderer: GLMapViewRenderer, private val imageManager: ImageManager) : LocationListener {
    private var userMovementImage: GLMapDrawable? = null
    private var userLocationImage: GLMapDrawable? = null
    private var accuracyCircle: GLMapDrawable? = null
    var isFollowLocationEnabled = false
        set(value) {
            field = value
            if (field) {
                val lastLocation = lastLocation
                if (lastLocation != null)
                    onLocationChanged(lastLocation)
            }
        }
    private var locationManager: LocationManager? = null
    private var lastLocation: Location? = null

    fun onDestroy() {
        val locationManager = locationManager
        if (locationManager != null) {
            this.locationManager = null
            locationManager.removeUpdates(this)
        }
    }

    fun initLocationManager(activity: Activity): Boolean {
        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION)
            != PackageManager.PERMISSION_GRANTED &&
            ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION)
            != PackageManager.PERMISSION_GRANTED
        ) return false
        var locationManager = locationManager
        if (locationManager == null) {
            try {
                // Setup get location service
                locationManager = activity.getSystemService(Context.LOCATION_SERVICE) as LocationManager

                // Configure criteria
                val criteria = Criteria()
                criteria.accuracy = Criteria.ACCURACY_FINE
                criteria.isSpeedRequired = true
                criteria.isAltitudeRequired = true
                criteria.isBearingRequired = true
                criteria.bearingAccuracy = Criteria.ACCURACY_HIGH

                // Find the best location that currently have all location providers
                var bestAccuracy = Float.MAX_VALUE
                var bestTime = Long.MAX_VALUE
                lastLocation = null
                val matchingProviders = locationManager.allProviders
                for (provider in matchingProviders) {
                    var location: Location? = null
                    try {
                        location = locationManager.getLastKnownLocation(provider)
                    } catch (e: Exception) {
                        Log.e("CurLocationHelper", e.localizedMessage ?: "No message")
                    }
                    if (location != null) {
                        val accuracy = location.accuracy
                        val time = location.time
                        if (time >= bestTime && accuracy < bestAccuracy) {
                            lastLocation = location
                            bestAccuracy = accuracy
                            bestTime = time
                        } else if (time < bestTime && bestAccuracy == Float.MAX_VALUE) {
                            lastLocation = location
                            bestTime = time
                            bestAccuracy = accuracy
                        }
                    }
                }
                // Update location to current best
                val lastLocation = lastLocation
                if (lastLocation != null)
                    onLocationChanged(lastLocation)
                // Request location updates
                locationManager.requestLocationUpdates(
                    MIN_TIME_BW_UPDATES,
                    MIN_DISTANCE_CHANGE_FOR_UPDATES.toFloat(),
                    criteria,
                    this,
                    activity.mainLooper
                )
            } catch (e: Exception) {
                locationManager = null
            } finally {
                this.locationManager = locationManager
            }
        }
        return true
    }

    override fun onLocationChanged(location: Location) {
        lastLocation = location
        val position = MapPoint.CreateFromGeoCoordinates(location.latitude, location.longitude)
        if (isFollowLocationEnabled)
            renderer.animate { it.flyToPoint(position) }

        // Create drawables if not exist and set initial positions.
        var userLocationImage = userLocationImage
        if (userLocationImage == null) {
            val locationImage = imageManager.open("circle_new.svg", 1f, 0)!!
            userLocationImage = GLMapDrawable(locationImage, 100)
            this.userLocationImage = userLocationImage
            userLocationImage.isHidden = true
            userLocationImage.setOffset(locationImage.width / 2, locationImage.height / 2)
            userLocationImage.position = position
            renderer.add(userLocationImage)
            locationImage.recycle()
        }

        var userMovementImage = userMovementImage
        if (userMovementImage == null) {
            val movementImage = imageManager.open("arrow_new.svg", 1f, 0)!!
            userMovementImage = GLMapDrawable(movementImage, 100)
            this.userMovementImage = userMovementImage
            userMovementImage.isHidden = true
            userMovementImage.setOffset(movementImage.width / 2, movementImage.height / 2)
            userMovementImage.isRotatesWithMap = true
            userMovementImage.position = position
            if (location.hasBearing()) userLocationImage.angle = -location.bearing
            renderer.add(userMovementImage)
            movementImage.recycle()
        }

        // Select what image to display
        if (location.hasBearing()) {
            userMovementImage.isHidden = false
            userLocationImage.isHidden = true
        } else {
            userLocationImage.isHidden = false
            userMovementImage.isHidden = true
        }

        // Calculate radius of accuracy circle
        val r = renderer.convertMetersToInternal(location.accuracy.toDouble()).toFloat()
        // If accuracy circle drawable not exits - create it and set initial position
        var accuracyCircle = accuracyCircle
        if (accuracyCircle == null) {
            val pointCount = 100
            // Use MapPoint to avoid distortions of circle
            val points = arrayOfNulls<MapPoint>(pointCount)
            for (i in 0 until pointCount) {
                val f = 2 * Math.PI * i / pointCount
                // If radius of circle will be 1 only 2 points will be in final geometry (after
                // douglas-peucker)
                points[i] = MapPoint(sin(f) * 2048, cos(f) * 2048)
            }
            val circle = GLMapVectorObject.createPolygon(arrayOf(points), null)
            accuracyCircle = GLMapDrawable(99)
            this.accuracyCircle = accuracyCircle
            accuracyCircle.setTransformMode(GLMapDrawable.TransformMode.Custom)
            accuracyCircle.position = position
            accuracyCircle.scale = r / 2048.0f.toDouble()
            accuracyCircle.setVectorObject(
                circle,
                GLMapVectorCascadeStyle.createStyle("area{layer:100; width:1pt; fill-color:#3D99FA26; color:#3D99FA26;}")!!,
                null
            )
            renderer.add(accuracyCircle)
        }
        renderer.animate {
            it.setTransition(GLMapAnimation.Linear)
            it.setDuration(1.0)
            userMovementImage.position = position
            userLocationImage.position = position
            accuracyCircle.position = position
            accuracyCircle.scale = r / 2048.0f.toDouble()
            if (location.hasBearing()) userLocationImage.angle = -location.bearing
        }
    }

    override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}
    override fun onProviderEnabled(provider: String) {}
    override fun onProviderDisabled(provider: String) {}

    companion object {
        private const val MIN_TIME_BW_UPDATES: Long = 1000 // 1 second
        private const val MIN_DISTANCE_CHANGE_FOR_UPDATES: Long = 1 // 1 meter
    }
}
